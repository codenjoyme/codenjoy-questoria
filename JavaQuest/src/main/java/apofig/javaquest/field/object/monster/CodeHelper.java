package apofig.javaquest.field.object.monster;

/**
 * User: sanja
 * Date: 02.07.13
 * Time: 20:00
 */
public interface CodeHelper {

    String getCode();

}
