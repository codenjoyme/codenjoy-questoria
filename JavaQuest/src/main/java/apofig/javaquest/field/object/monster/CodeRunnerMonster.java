package apofig.javaquest.field.object.monster;

import apofig.compiler.JavaCompiler;
import apofig.javaquest.field.object.monster.test.TestResult;
import apofig.javaquest.field.object.monster.test.TestSuite;

public abstract class CodeRunnerMonster extends Monster implements MonsterTest {

    public final static String HELP = "Попробуй еще раз!";
    public final static String LEAVE = "Просто так ты не уйдешь!";
    public static final String ANSWER = "4E0A";

    protected MonsterTest test;
    protected TestSuite testSuite = new TestSuite();

    public CodeRunnerMonster(String question, String signature, int weight) {
        super(question, ANSWER, HELP, LEAVE, signature, weight);
        this.test = this;
    }

    @Override
    public void answer(String code) {
        if (code.equals(ANSWER)) { // TODO подумать и упростить этот костыль с 4E0A/OK и другими константами
            super.answer(code);
            return;
        }
        try {
            MethodRunner method = compile(code);
            Object[][] testData = getTestData();
            for (Object[] data : testData) {
                TestResult testResult = test.test(data, method);
                testSuite.add(testResult);
                if (testSuite.isEnough()) break;
            }
        } catch (Exception e) {
            testSuite.add(new TestResult(e));
        }
        processResults(testSuite.getResults());
        testSuite.clear();
    }

    protected abstract Object[][] getTestData();

    protected void processResults(String message) {
        if (message.equals("OK")) {
            super.answer("4E0A");
        } else {
            messenger.say(message);
            messenger.say(help);
        }
    }

    private MethodRunner compile(String code) {
        JavaCompiler compiler = new JavaCompiler();
        return compiler.getMethod(code);
    }

    public TestSuite getTestSuite() {
        return testSuite;
    }
}
