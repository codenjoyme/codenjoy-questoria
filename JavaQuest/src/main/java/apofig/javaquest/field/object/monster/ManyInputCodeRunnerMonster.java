package apofig.javaquest.field.object.monster;

import apofig.javaquest.field.object.monster.test.TestResult;

public abstract class ManyInputCodeRunnerMonster extends WSCodeRunnerMonster {

    public ManyInputCodeRunnerMonster(String question, String signature, int weight) {
        super(question, signature, weight);
    }

    @Override
    public TestResult test(Object[] test, MethodRunner method) {
        try {
            Object actual = method.run(test);
            String expected = method(test);
            return new TestResult(test, actual, expected);
        } catch (Exception e) {
            return new TestResult(test, e);
        }
    }

    protected abstract String method(Object...inputs);

}
