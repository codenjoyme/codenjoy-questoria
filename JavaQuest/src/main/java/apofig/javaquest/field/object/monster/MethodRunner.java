package apofig.javaquest.field.object.monster;

/**
 * User: oleksandr.baglai
 * Date: 2/8/13
 * Time: 1:40 PM
 */
public interface MethodRunner {
    Object run(Object... objects);
}
