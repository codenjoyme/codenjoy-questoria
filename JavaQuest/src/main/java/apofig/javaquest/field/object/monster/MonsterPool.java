package apofig.javaquest.field.object.monster;

public interface MonsterPool {
    Monster next();
}
