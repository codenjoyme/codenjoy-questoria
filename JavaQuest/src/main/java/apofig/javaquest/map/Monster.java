package apofig.javaquest.map;

/**
 * User: oleksandr.baglai
 * Date: 1/19/13
 * Time: 8:30 PM
 */
public class Monster implements Something {

    private final OnKill onKill;

    public Monster(OnKill onKill) {
        this.onKill = onKill;
    }

    @Override
    public String answer(String message) {
        if (message.equals("die!")) {
            onKill.doit();
            return "Monster: yOU @#& Ki$%@&^ll me $!@!";
        }
        return "Monster: I'll kill you!";
    }

    @Override
    public boolean iCanLeave() {
        return false;
    }

    @Override
    public String askMe() {
        return "Monster: Fight with me!";
    }

    @Override
    public boolean iCanUse() {
        return false;
    }
}
