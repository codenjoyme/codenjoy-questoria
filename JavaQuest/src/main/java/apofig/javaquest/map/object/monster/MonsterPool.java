package apofig.javaquest.map.object.monster;

/**
 * User: oleksandr.baglai
 * Date: 2/5/13
 * Time: 10:17 PM
 */
public interface MonsterPool {
    Monster next();
}
