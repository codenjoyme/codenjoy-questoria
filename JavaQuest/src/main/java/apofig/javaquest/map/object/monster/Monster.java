package apofig.javaquest.map.object.monster;

import apofig.javaquest.map.Action;
import apofig.javaquest.map.Dieble;
import apofig.javaquest.map.Player;
import apofig.javaquest.map.object.*;

/**
 * User: oleksandr.baglai
 * Date: 1/19/13
 * Time: 8:30 PM
 */
public class Monster extends TalkingObject implements Something, CodeHelper, Dieble, SetWorld {

    private String question;
    private String answer;
    protected String help;
    private String leave;
    private String signature;
    private Action onKill;
    private World world;
    private int weight;

    public Monster(String question, String answer,
                   String help, String leave,
                   String signature, int weight)
    {
        this.question = question;
        this.answer = answer;
        this.help = help;
        this.leave = leave;
        this.signature = signature;
        this.weight = weight;
    }

    @Override
    public void answer(String message) {
        if (message.equals(answer)) {
            messenger.say("тЫ @#& Уб$%@&^ил ме:ня $!@!");
            Something gold = leaveAfter();
            gold.ask();
        } else {
            messenger.say(help);
        }
    }

    @Override
    public boolean canLeave() {
        return false;
    }

    @Override
    public void ask() {
        messenger.say(question);
    }

    @Override
    public boolean canUse() {
        return false;
    }

    @Override
    public Something leaveAfter() {
        return world.make('$');
    }

    @Override
    public char symbol() {
        return '@';
    }

    @Override
    public void getBy(Player info) {
        // do nothing
    }

    @Override
    public void tryToLeave() {
        messenger.say(leave);
    }

    public void onKill(Action action) {
        onKill = action;
    }

    @Override
    public void die() {
        if (onKill != null) {
            onKill.act(this);
        }
    }

    @Override
    public String getCode() {
        return signature;
    }

    @Override
    public void setWorld(World world) {
        this.world = world;
    }

    public int getWeight() {
        return weight;
    }

    @Override
    public String toString() {
        return String.format("[Monster: %s with weight: %s]", this.getClass().getSimpleName(), weight);
    }
}
