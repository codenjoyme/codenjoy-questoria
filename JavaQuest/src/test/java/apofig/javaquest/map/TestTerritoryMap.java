package apofig.javaquest.map;

import apofig.javaquest.map.object.monster.Monster;
import apofig.javaquest.map.object.monster.MonsterPool;
import org.approvaltests.Approvals;
import org.junit.Before;
import org.junit.Test;

import static apofig.javaquest.map.Messages.withoutSeparator;
import static junit.framework.Assert.assertEquals;
import static org.mockito.Mockito.mock;

/**
 * User: oleksandr.baglai
 * Date: 1/19/13
 * Time: 2:08 PM
 */
public class TestTerritoryMap {

    private JavaQuest game;
    private TerritoryMap map;
    private Joystick joystick;

    public int getWidth() {
        return 100;
    }

    public int getHeight() {
        return 100;
    }

    public int getViewAreaSize() {
        return 13;
    }

    public int getMonsterX() {
        return 40;
    }

    public int getMonsterY() {
        return 22;
    }

    public int getWallX() {
        return 22;
    }

    public int getWallY() {
        return 10;
    }

    public RectangleMap getMapLoader() {
        return new SquareMap(getWidth());
    }

    @Before
    public void init() {
        final RectangleMap mapLoader = getMapLoader();
        mapLoader.setMonster(getMonsterX(), getMonsterY());
        mapLoader.setWall(getWallX(), getWallY());

        Settings settings = new Settings() {
            @Override
            public int getViewAreaSize() {
                return TestTerritoryMap.this.getViewAreaSize();
            }

            @Override
            public MapLoader getMapLoader() {
                return mapLoader;
            }

            @Override
            public MonsterPool getMonsters() {
                return new MonsterPool() {
                    @Override
                    public Monster next() {

                        return new Monster("Сразись со мной!",
                                "die!",
                                "Я убью тебя!",
                                "Никуда ты не уйдешь!",
                                "немногоКода('для подсказки');", null);

                    }
                };
            }
        };
        game = new JavaQuest(settings);
        map = game.getTerritoryMap();
        joystick = game.getPlayer();
    }

    @Test
    public void testIAmOnMap() throws Exception {
        verifyMap();
    }

    private void verifyMap() throws Exception {
        Approvals.verify(getMap());
    }

    private String getMap() {
        StringOutputStream out = new StringOutputStream();
        map.printNearMe(out);
        return out.getResult();
    }

    @Test
    public void testIGoRight() throws Exception {
        moveRight();

        verifyMap();
    }

    @Test
    public void testIGoLeft() throws Exception {
        moveLeft();

        verifyMap();
    }

    @Test
    public void testIGoUp() throws Exception {
        moveUp();

        verifyMap();
    }

    @Test
    public void testIGoDownTwice() throws Exception {
        moveDown();
        moveDown();

        verifyMap();
    }

    @Test
    public void testCheckGoToBoardLeftDown() throws Exception {
        moveTo(5, 5);

        verifyMap();
    }

    @Test
    public void testTryToOpenViewOnBoardLeftDown() throws Exception {
        moveTo(1, 1);

        verifyMap();
    }

    @Test
    public void testCheckGoToBoardUpRight() throws Exception {
        if (getHeight() > getWidth()) { // fix for TestTerritoryRectangleMap test
            moveTo(20, 20 + (getHeight() - getWidth()));
        }
        moveTo(getWidth() - 6, getHeight() - 6);

        verifyMap();
    }

    @Test
    public void testTryToOpenViewOnBoardUpRight() throws Exception {
        if (getHeight() > getWidth()) { // fix for TestTerritoryRectangleMap test
            moveTo(20, 20 + (getHeight() - getWidth()));
        }
        moveTo(getWidth() - 2, getHeight() - 2);

        verifyMap();
    }

    @Test
    public void testICantGoOnBoardDown() throws Exception {
        moveTo(20, 0);
        moveDown();

        verifyXY(20, 0);
        assertMessage("Wall: Пожалуйста, остановись!");
        verifyMap();
    }

    @Test
    public void testICantGoOnBoardUp() throws Exception {
        moveTo(20, getHeight() - 1);
        moveUp();

        verifyXY(20, getHeight() - 1);
        assertMessage("Wall: Пожалуйста, остановись!");
        verifyMap();
    }

    @Test
    public void testICantGoOnBoardLeft() throws Exception {
        moveTo(0, 20);
        moveLeft();

        verifyXY(0, 20);
        assertMessage("Wall: Пожалуйста, остановись!");
        verifyMap();
    }

    private void verifyXY(int x, int y) {
        assertEquals(x, map.getX());
        assertEquals(y, map.getY());
    }

    @Test
    public void testICantGoOnBoardRight() throws Exception {
        moveTo(getWidth() - 1, 20);
        moveRight();

        verifyXY(getWidth() - 1, 20);
        assertMessage("Wall: Пожалуйста, остановись!");
        verifyMap();
    }

    @Test
    public void testWhenMoveICanFindMonster() throws Exception {
        moveTo(getMonsterX() - 3, getMonsterY());

        verifyMap();
    }

    private void moveTo(int x, int y) throws Exception {
        int count = 0;
        while (Math.abs(map.getX() - x) != 0 || Math.abs(map.getY() - y) != 0) {
            if (count++ > 1000) {
                throw new RuntimeException(String.format(
                        "Я не могу пройти сюда! My x=%s and y=%s, and view:\n%s",
                        map.getX(), map.getY(), map.getViewArea()));
            }
            if (map.getY() < y) {
                moveUp();
            } else if (map.getY() > y) {
                moveDown();
            }
            if (map.getX() < x) {
                moveRight();
            } else if (map.getX() > x) {
                moveLeft();
            }
        }
        verifyXY(x, y);
    }

    private void moveLeft() {
        joystick.moveLeft();
        game.tick();
    }

    private void moveUp() {
        joystick.moveUp();
        game.tick();
    }

    private void moveDown() {
        joystick.moveDown();
        game.tick();
    }

    @Test
    public void testWhenITalkWithMonster() throws Exception {
        moveTo(getMonsterX() - 1, getMonsterY());

        assertMessage("Monster: Сразись со мной!");

        assertEquals("немногоКода('для подсказки');",
                game.getCodeHelper().getCode());

        verifyMap();
    }

    @Test
    public void shouldNoMoveWhenITalkWithMonster() throws Exception {
        moveTo(getMonsterX() - 1, getMonsterY());
        moveRight();

        verifyXY(getMonsterX() - 1, getMonsterY());
        assertMessage("Monster: Сразись со мной!\n" +
                "Monster: Никуда ты не уйдешь!");
        verifyMap();
    }

    @Test
    public void shouldKillMonsterWhenAttack() throws Exception {
        moveTo(getMonsterX() - 1, getMonsterY());
        attack("die!");

        assertMessage("Monster: Сразись со мной!\n" +
                "You: die!\n" +
                "Monster: тЫ @#& Уб$%@&^ил ме:ня $!@!\n" +
                "Gold: Привет, я - 10$");
        verifyMap();
    }

    @Test
    public void shouldMonsterStillAliveWhenBadAttack() throws Exception {
        moveTo(getMonsterX() - 1, getMonsterY());
        attack("No!!!");

        assertMessage("Monster: Сразись со мной!\n" +
                "You: No!!!\n" +
                "Monster: Я убью тебя!");
        verifyMap();
    }

    private void attack(String message) {
        joystick.attack(message);
        game.tick();
    }

    @Test
    public void shouldMonsterStillAliveWhenBadAttackTwice() throws Exception {
        moveTo(getMonsterX() - 1, getMonsterY());
        attack("No!!!");
        attack("Nooooo!!!");

        assertMessage("Monster: Сразись со мной!\n" +
                "You: No!!!\n" +
                "Monster: Я убью тебя!\n" +
                "You: Nooooo!!!\n" +
                "Monster: Я убью тебя!");
        verifyMap();
    }

    @Test
    public void shouldNoMoveWhenITryToGoOnWall() throws Exception {
        moveTo(getWallX() - 1, getWallY());
        moveRight();

        verifyXY(getWallX() - 1, getWallY());
        assertMessage("Wall: Пожалуйста, остановись!");
        verifyMap();
    }

    private void moveRight() {
        joystick.moveRight();
        game.tick();
    }

    @Test
    public void shouldGetGoldAfterMonsterDie() throws Exception {
        moveTo(getMonsterX() - 1, getMonsterY());
        assertInfo("Уровень:0 Опыт:0 Здоровье:100 Золото:0");

        attack("die!");
        moveRight();

        assertMessage("Monster: Сразись со мной!\n" +
                "You: die!\n" +
                "Monster: тЫ @#& Уб$%@&^ил ме:ня $!@!\n" +
                "Gold: Привет, я - 10$\n" +
                "Gold: Ну и ладно! Достанусь кому-то другому!!\n" +
                "Gold: Привет, я - 10$\n" +
                "Gold: Ты подобрал меня! Спасибо!!");
        verifyMap();

        assertInfo("Уровень:0 Опыт:0 Здоровье:100 Золото:10");
    }

    private void assertInfo(String info) {
        assertEquals(info, game.getPlayerInfo().toString());
    }

    private void assertMessage(String message) {
        assertEquals(message, withoutSeparator(game.getMessage()));
    }

    @Test
    public void shouldLeaveGoldAfterMonsterDie() throws Exception {
        moveTo(getMonsterX() - 1, getMonsterY());
        attack("die!");

        assertMessage("Monster: Сразись со мной!\n" +
                "You: die!\n" +
                "Monster: тЫ @#& Уб$%@&^ил ме:ня $!@!\n" +
                "Gold: Привет, я - 10$");

        verifyMap();
    }

    @Test
    public void shouldNoRepeatMessageAgain() throws Exception {
        moveTo(getMonsterX() - 1, getMonsterY());
        attack("die!");
        moveUp();

        assertMessage("Monster: Сразись со мной!\n" +
                "You: die!\n" +
                "Monster: тЫ @#& Уб$%@&^ил ме:ня $!@!\n" +
                "Gold: Привет, я - 10$");

        verifyMap();
    }

    @Test
    public void shouldNoKillGold() throws Exception {
        moveTo(getMonsterX() - 1, getMonsterY());
        attack("die!");
        attack("Gold die!");

        assertMessage("Monster: Сразись со мной!\n" +
                "You: die!\n" +
                "Monster: тЫ @#& Уб$%@&^ил ме:ня $!@!\n" +
                "Gold: Привет, я - 10$\n" +
                "You: Gold die!\n" +
                "Gold: Ты не можешь делать это со мной!");
        verifyMap();
    }

    @Test
    public void shouldSomeMessageAfterGetGold() throws Exception {
        moveTo(getMonsterX() - 1, getMonsterY());
        attack("die!");
        moveRight();
        moveRight();

        assertMessage("Monster: Сразись со мной!\n" +
                "You: die!\n" +
                "Monster: тЫ @#& Уб$%@&^ил ме:ня $!@!\n" +
                "Gold: Привет, я - 10$\n" +
                "Gold: Ну и ладно! Достанусь кому-то другому!!\n" +
                "Gold: Привет, я - 10$\n" +
                "Gold: Ты подобрал меня! Спасибо!!");
        verifyMap();
    }

    @Test
    public void shouldGoAwayFromGold() throws Exception {
        moveTo(getMonsterX() - 1, getMonsterY());
        attack("die!");
        moveLeft();
        moveLeft();

        assertMessage("Monster: Сразись со мной!\n" +
                "You: die!\n" +
                "Monster: тЫ @#& Уб$%@&^ил ме:ня $!@!\n" +
                "Gold: Привет, я - 10$\n" +
                "Gold: Ну и ладно! Достанусь кому-то другому!!");
        verifyMap();
    }

}
