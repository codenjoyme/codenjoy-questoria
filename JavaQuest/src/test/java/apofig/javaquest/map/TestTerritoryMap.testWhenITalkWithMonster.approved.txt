╔══════════════════════════╗
║? ? ? ? ? ? ? ? ? ? ? ? ? ║
║                ? ? ? ? ? ║
║                    ? ? ? ║
║                      ? ? ║
║                      ? ? ║
║                        ? ║
║            I @         ? ║
║                        ? ║
║                      ? ? ║
║                      ? ? ║
║                    ? ? ? ║
║                ? ? ? ? ? ║
║? ? ? ? ? ? ? ? ? ? ? ? ? ║
╚══════════════════════════╝
Monster: Fight with me!